### ISSUE TITLE
How to use configuration file

### ISSUE THREAD

**Issue by iamnewton** [markdownlint/markdownlint#67]
This isn't really a bug, but more of an inquiry.  How do I use the `.mdlrc`?  I understand that I can just list out all of the rules that I want to use, but some of the rules have other options that allow enforcement of particular styles, such as [this one](https://github.com/mivok/markdownlint/blob/master/docs/RULES.md#md003---header-style).  

Note: An example file would go a long way in your documentation.

**Comment by ndarville** [comment]
Hey, @chrisopedia. I was wondering the exact same thing as you just now, until I went through the `docs/` folder by myself and came upon [this](https://github.com/mivok/markdownlint/blob/master/docs/configuration.md) configuration guide.

If you want an explicit set of rules, create an `.mdlrc` and fill in something like:

```
rules "MD001", "MD005", "MD006", "MD011", "MD014", "MD018", "MD019", "MD020", "MD021", "MD024", "MD025", "MD026", "MD027", "MD028"
```

When you want to lint your Markdown, run

``` sh
mdl README.md -c .mdlrc
```

This works for me.

If you want an example, you can use my repo:

```
git clone git@github.com:ndarville.com/style.git
cd style/markdown
# Add some errors in markdown/README.md manually
mdl README.md -c .mdlrc
```

You have to tell `mdl` of your config file explicitly; otherwise, it will ignore it.

**Comment by mivok** [comment]
Hi Chris,

You are correct that in .mdlrc you just list out the rules you want to enforce. This is the same as the -r option on the command line (everything in .mdlrc corresponds to a command line option). If you want to configure a specific style such as in MD003, this is done by creating a style file. A style file provides more powerful filtering on which rules to enable as well as allowing you to configure them. An example style file is here: https://github.com/mivok/markdownlint/blob/master/lib/mdl/styles/cirosantilli.rb and you can see some documentation on creating them at https://github.com/mivok/markdownlint/blob/master/docs/creating_styles.md.

Note: a style file doesn't have to be located in the styles directory like the example given, you can specify the path to the style file instead of just a name with the -s option on the command line or with the style option in .mdlrc.

**Comment by iamnewton** [comment]
So if you were to create a `.mdlrc` with the listing of rules you wanted, would you create a style.rb for more filtering or would you use it as the rules file?  I'm only confused because the doc for creating styles and the example you provided have the 'all' listed in there which makes me feel like I'll be duplicating my rules file within a style file.

**Comment by iamnewton** [comment]
Getting an error with trying the `.mdlrc` and `style.rb`.

markdown.style.rb

```
rule 'MD003', :style => :atx
rule 'MD004', :style => :dash
rule 'MD007', :indent => 4
rule 'MD029', :style => :ordered
```

.mdlrc

```
rules "MD001", // header levels increment by 1
      "MD002", // first header should be level 1
      "MD003", // header style - atx
      "MD004", // unordered list style - asterick
      "MD005", // don't allow inconsistent indentation for list items
      "MD006", // start bullets at beginning of line
      "MD007", // unordered list indentation should be 4 spaces
      "MD009", // trailing spaces not allowed
      "MD010", // hard tabs not allowed
      "MD011", // alert on reversed link syntax
      "MD012", // alert on multiple consecutive blank lines
      "MD013", // line length should be no more than 80 characters
      "MD014", // dollars signs not before commands w/o showing output
      "MD018", // must have space after header style
      "MD019", // no multiple spaces after header style
      "MD022", // headers should be surrounded by blank lines
      "MD023", // headers must start at the beginning of the line
      "MD024", // no multiple headers with same content
      "MD025", // only 1 level 1 header
      "MD026", // no trailing punctuation in headers
      "MD027", // no multiple spaces after blockquote symbol
      "MD028", // no blank lines within blockquote
      "MD029", // ordered list item prefix, shoulded be ordered
      "MD030", // spaces after list markers
      "MD031", // fenced code blocks need line surrounding
      "MD032" // lists should be surrounded by line
style "/usr/local/opt/dotfiles/conf/misc/markdown.style.rb"
```

Error

```
/Library/Ruby/Gems/2.0.0/gems/mixlib-config-2.1.0/lib/mixlib/config.rb:47:in `instance_eval': /Users/cnewton/.mdlrc|1 error| syntax error, unexpected tIDENTIFIER, expecting end-of-input (SyntaxError)
```

**Comment by mivok** [comment]
To fix the syntax error, change `//` to `#`. `//` doesn't work as a comment character in ruby (and .mdlrc).

However, to answer your earlier question about .mdlrc vs style files, here's an explanation of the reasoning behind it. The short version is that I think you actually just want to use a style file and not bother with any rules settings in .mdlrc.
- The `--rules` option on the command line is intended to allow you to quickly filter out which rules are enabled for a given run.
- The configuration in .mdlrc is intended to allow you to have certain command line options apply by default (such as the kramdown warnings, or which style to use).
- Style files are intended to allow you to configure your personal markdown style (or use someone else's) that you want to conform to. This includes enabling certain rules, but also configuring their behavior (such as 'all list items should begin with a minus sign'). They provide more flexible selection of which rules should be enabled/disabled than is possible in .mdlrc.

The example style files all start by enabling all rules, and configuring their behavior, because they assume all rules should be enabled. In your case however, you could do something like this for your style file, which only enables certain rules:

```
rule "MD001" # header levels increment by 1
rule "MD002" # first header should be level 1
rule "MD003", :style => :atx # header style - atx
rule "MD004", :style => :dash # unordered list style - dash
rule "MD005" # don't allow inconsistent indentation for list items
rule "MD006" # start bullets at beginning of line
rule "MD007", :indent => 4 # unordered list indentation should be 4 spaces
rule "MD009" # trailing spaces not allowed
rule "MD010" # hard tabs not allowed
rule "MD011" # alert on reversed link syntax
rule "MD012" # alert on multiple consecutive blank lines
rule "MD013" # line length should be no more than 80 characters
rule "MD014" # dollars signs not before commands w/o showing output
rule "MD018" # must have space after header style
rule "MD019" # no multiple spaces after header style
rule "MD022" # headers should be surrounded by blank lines
rule "MD023" # headers must start at the beginning of the line
rule "MD024" # no multiple headers with same content
rule "MD025" # only 1 level 1 header
rule "MD026" # no trailing punctuation in headers
rule "MD027" # no multiple spaces after blockquote symbol
rule "MD028" # no blank lines within blockquote
rule "MD029", :style => :ordered # ordered list item prefix, shoulded be ordered
rule "MD030" # spaces after list markers
rule "MD031" # fenced code blocks need line surrounding
rule "MD032" # lists should be surrounded by line
```

And then just have the following line in your .mdlrc:

```
 style "/usr/local/opt/dotfiles/conf/misc/markdown.style.rb"
```

**Comment by iamnewton** [comment]
Perfect.  Thanks so much.  Love this project by the way.

**Comment by mivok** [comment]
Thanks, and you're welcome!

**Comment by devth** [comment]
I still don't know how to use `.mdlrc`. You absolutely need a sample file!
- https://github.com/mivok/markdownlint/blob/master/lib/mdl/styles/cirosantilli.rb shows how to specify 4 options. What about the others?
- `rule 'MD007', :indent => 4` is meaningless unless you look up RULES.md

How do I specify inline html? Is it:

```
rule 'MD033', :html
```

or

```
rule 'MD033', :html => true
```

and why can't it just be:

```
:allow_html => true
```

or 

```
allow_html: true
```

Great project :+1:

**Comment by GarthDB** [comment]
I super agree @devth a sample file would be awesome.

**Comment by tonglil** [comment]
How do you tell MDL to ignore a rule?

**Comment by ndarville** [comment]
@tonglil

This is what my `.markdownlintrc` looks like`:

```json
{
      "default": false,
      "MD001": true,
      "MD005": true,
      "MD006": true,
      "MD011": true,
      "MD014": true,
      "MD018": true,
      "MD019": true,
      "MD020": true,
      "MD021": true,
      "MD025": true,
      "MD027": true,
      "MD028": true,
      "MD031": true,
      "MD034": true,
      "MD037": true,
      "MD038": true
}
```

And this is the command I run with my linter:

```sh
markdownlint --config node_modules/style/markdown/.markdownlintrc _posts/*.markdown
```

**Comment by mloskot** [comment]
https://github.com/markdownlint/markdownlint/issues/67#issuecomment-277440662 suggestion does not work forme:

`cat .markdownlint.json`
```
{
    "MD013": false
}
```

then

```
mdl --config .markdownlint.json README.md

README.md:17: MD013 Line length
README.md:18: MD013 Line length
README.md:43: MD032 Lists should be surrounded by blank lines
README.md:47: MD032 Lists should be surrounded by blank lines
README.md:55: MD034 Bare URL used
```

and it still complains about MD013.

Please, an example of `.json/rc` file would be priceless.

**Comment by neenjaw** [comment]
@mloskot - Just in case this still isn't working for you, this seems to be working for me now with this file: 

```json
{
    "line-length": false,
    "first-header-h1": false
}
```
the key is the error's name with dashes (in vs code, when the error is automatically linted it, you can find it in the popover in the format: "MD002/first-header-h1"

**Comment by DavidAnson** [comment]
When you say VS Code, I think you mean https://marketplace.visualstudio.com/items?itemName=DavidAnson.vscode-markdownlint which is powered by a Node.js version of this code, https://github.com/DavidAnson/markdownlint

It tries to be feature-compatible, but if you are hitting a bug (in either), that problematic behavior may not be present in the other implementation.

**Comment by neenjaw** [comment]
It works well for me with the JSON config I posted, but sorry for mixing up the packages. Thanks David!

**Comment by fauust** [comment]
Hi!
I know this is closed but I do not understand what is the cost to put an example in your documentation as requested [many](https://github.com/markdownlint/markdownlint/issues/67#issuecomment-331579326), [many](https://github.com/markdownlint/markdownlint/issues/67#issuecomment-195164828), [many](https://github.com/markdownlint/markdownlint/issues/67#issue-55995077) times?

I can't understand what I am doing wrong:
```bash
❯ cat ~/.mdlrc                                                                                             
exclude_rule 'MD009'
❯ mdl README.md -c ~/.mdlrc                                                                              
README.md:6: MD009 Trailing spaces

A detailed description of the rules is available at https://github.com/markdownlint/markdownlint/blob/master/docs/RULES.md
```

**Comment by fauust** [comment]
Ok, I found it thanks to https://github.com/30-seconds/30-seconds-of-code/blob/master/.mdlrc hope this could help others.

The following disables `MD013` and configure `MD009` to accept two trailing spaces (markdown line breaks):
```bash
❯ cat .mdlrc                                                                                             
rules "~MD013"
style "/home/faust/.mdlrc.style.rb"
❯ cat .mdlrc.style.rb                                                                                    
all
rule 'MD009', :br_spaces => 2
```
But again, please consider adding samples in the documentation...

**Comment by mkesper** [comment]
Why can't style settings be inline in .mdlrc? This really clutters projects with unnecessary files.
